# Finding the most visited Venues of my city Bengaluru

### Need to suggest my friend in which area there are frequent visitors to Indian Restaurant so that he can open expand his restaurant chain and open the new cafe

#### importing the details of Zone, Area and Pincode 

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('BangaloreDetails.csv')

In [2]:
df=df.replace("Not assigned", np.nan)
df.dropna(subset=['Zone'],axis=0, inplace=True)
df.head()


Zone               Area
0  Central    cantonment area
1  Central             Domlur
2  Central        Indiranagar
3  Central  Jeevanbheemanagar
4  Central        Malleswaram

### importing the Libraries

In [3]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

print('Libraries imported.')



Solving environment: - 
The environment is inconsistent, please check the package plan carefully
The following packages are causing the inconsistency:

  - defaults/linux-64::anaconda==5.3.1=py37_0
  - defaults/linux-64::astropy==3.0.4=py37h14c3975_0
  - defaults/linux-64::bkcharts==0.2=py37_0
  - defaults/linux-64::blaze==0.11.3=py37_0
  - defaults/linux-64::bokeh==0.13.0=py37_0
  - defaults/linux-64::bottleneck==1.2.1=py37h035aef0_1
  - defaults/linux-64::dask==0.19.1=py37_0
  - defaults/linux-64::datashape==0.5.4=py37_1
  - defaults/linux-64::mkl-service==1.1.2=py37h90e4bf4_5
  - defaults/linux-64::numba==0.39.0=py37h04863e7_0
  - defaults/linux-64::numexpr==2.6.8=py37hd89afb7_0
  - defaults/linux-64::odo==0.5.1=py37_0
  - defaults/linux-64::pytables==3.4.4=py37ha205bf6_0
  - defaults/linux-64::pytest-arraydiff==0.2=py37h39e3cac_0
  - defaults/linux-64::pytest-astropy==0.4.0=py37_0
  - defaults/linux-64::pytest-doctestplus==0.1.3=py37_0
  - defaults/linux-64::pywavelets==1.0.0=py37h

### Fetch the Latitude and Longitude based on Pincode

In [4]:
result_latitude = []
result_longitude = []
for index, row in df.iterrows():
    print(row.Area)
    address = row.Area
    try:
        geolocator = Nominatim(user_agent="ny_explorer")
        location = geolocator.geocode('{}, Bangalore, Karnataka'.format(address))
        latitude = location.latitude
        result_latitude.append(latitude)
        longitude = location.longitude
        result_longitude.append(longitude)
        print('The geograpical coordinate of {} are {}, {}.'.format(address,latitude, longitude))
        print (location)
        
    except:
        
        pass

df["Latitude"] = pd.Series(result_latitude)
df["Longitude"] = pd.Series(result_longitude)

    

cantonment area
The geograpical coordinate of Toronto are 13.019567, 77.5095888861308.
Peenya Industrial Area, Dasarahalli Zone, Bengaluru, Bangalore Urban, Karnataka, 560058, India
Domlur
Indiranagar
The geograpical coordinate of Toronto are 12.9732913, 77.6404672.
Indiranagar, East Zone, Bengaluru, Bangalore Urban, Karnataka, 560038, India
Jeevanbheemanagar
Malleswaram
The geograpical coordinate of Toronto are 13.0163411, 77.5586641823841.
Malleswaram, West Zone, Bengaluru, Bangalore Urban, Karnataka, 560055, India
Pete area
Rajajinagar
The geograpical coordinate of Toronto are 12.9882338, 77.554883.
Rajaji Nagar, West Zone, Bengaluru, Bangalore Urban, Karnataka, 560010, India
Sadashivanagar
The geograpical coordinate of Toronto are 13.0077079, 77.5795893.
13th Cross Sadashivanagar, Aramane Nagara Ward, West Zone, Bengaluru, Bangalore Urban, Karnataka, 560080, India
Seshadripuram
The geograpical coordinate of Toronto are 12.9931876, 77.5753419.
Seshadripuram, West Zone, Bengaluru, Ba

In [12]:
df = df.dropna(how='any')

In [13]:
df

Zone                 Area   Latitude  Longitude
0            Central      cantonment area  13.019567  77.509589
1            Central               Domlur  12.973291  77.640467
2            Central          Indiranagar  13.016341  77.558664
3            Central    Jeevanbheemanagar  12.988234  77.554883
4            Central          Malleswaram  13.007708  77.579589
5            Central            Pete area  12.993188  77.575342
6            Central          Rajajinagar  12.986391  77.607542
7            Central       Sadashivanagar  12.977879  77.624670
8            Central        Seshadripuram  12.988721  77.585169
9            Central         Shivajinagar  12.979120  77.591300
10           Central               Ulsoor  12.991903  77.716201
11           Central        Vasanth Nagar  13.007516  77.695935
12           Eastern            Bellandur  13.067116  77.320126
13           Eastern       CV Raman Nagar  12.955257  77.698416
14           Eastern                Hoodi  12.940615  77.746994
15           Eastern     Krishnarajapuram  12.969637  77.749745
16           Eastern         Mahadevapura  13.014162  77.651854
17           Eastern         Marathahalli  13.035870  77.632360
18           Eastern              Varthur  13.027331  77.660151
19           Eastern           Whitefield  13.007964  77.636670
20     North-Eastern           Banaswadi   13.011768  77.627829
21     North-Eastern          HBR Layout   13.012022  77.677782
22     North-Eastern            Horamavu   13.038218  77.591900
23     North-Eastern        Kammanahalli   13.046453  77.548380
24     North-Eastern      Lingarajapuram   13.032888  77.557374
25     North-Eastern    Ramamurthy Nagar   13.032942  77.527325
26          Northern              Hebbal   13.076641  77.557732
27          Northern           Jalahalli   13.100698  77.596345
28          Northern           Mathikere   13.015958  77.545116
29          Northern              Peenya   13.345732  77.433335
30          Northern         R. T. Nagar   12.816244  77.691611
31          Northern      Vidyaranyapura   12.915177  77.610282
32          Northern           Yelahanka   12.979120  77.591300
33          Northern        Yeshwanthpur   12.911623  77.638862
34     South-Eastern        Bommanahalli   13.292399  77.751926
35     South-Eastern         Bommasandra   12.921227  77.617951
36     South-Eastern          BTM Layout   12.915221  77.573598
37     South-Eastern     Electronic City   12.941726  77.575502
38     South-Eastern          HSR Layout   12.940144  77.544500
39     South-Eastern         Koramangala   12.909492  77.596739
40     South-Eastern            Madiwala   12.929273  77.582423
41          Southern        Banashankari   12.906817  77.563525
42          Southern        Basavanagudi   12.918160  77.557674
43          Southern           Girinagar   12.905568  77.545544
44          Southern         J. P. Nagar   12.858081  77.558071
45          Southern           Jayanagar   12.878470  77.604179
46          Southern  Kumaraswamy Layout   12.863389  77.613011
47          Southern     Padmanabhanagar   13.181988  77.243144
48          Southern          Uttarahalli  12.877349  77.602803
49  Southern suburbs          Anjanapura   12.874255  77.583725
50  Southern suburbs             Arekere   12.986475  77.538571
51  Southern suburbs               Begur   12.982362  77.522638
52  Southern suburbs           Gottigere   12.919940  77.483349
53  Southern suburbs            Hulimavu   13.011894  77.543858
54  Southern suburbs              Kothnur  12.965101  77.507863
55           Western   Basaveshwaranagar   13.010406  77.537803
56           Western       Kamakshipalya   12.941325  77.521212
57           Western             Kengeri   12.927428  77.515372
58           Western  Mahalakshmi Layout   12.971889  77.545789

### Plot the Map of Bengaluru

In [14]:
map_bengaluru = folium.Map(location=[latitude, longitude], zoom_start=10)

for lat, lng, zone, area in zip(df['Latitude'], df['Longitude'], df['Zone'], df['Area']):
    label = '{}, {}'.format(area, zone)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_bengaluru)  
    
map_bengaluru

### Declare the FourSquare access details

In [15]:
CLIENT_ID = '**********************' # your Foursquare ID
CLIENT_SECRET = '*******************************' # your Foursquare Secret
VERSION = '20192904' # Foursquare API version

In [16]:
neighborhood_latitude = df.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df.loc[0, 'Area'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))



Latitude and longitude values of cantonment area are 13.019567, 77.5095888861308.


### Fetch the Top 10 Venus in each area

In [17]:
LIMIT=10

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
        print(lat)
        print(lng)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
        print (url)    
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

df2 = df.apply(pd.to_numeric, errors='coerce').fillna(77.5095888861308, downcast='infer')
df2.head()

try_venues = getNearbyVenues(names=df['Area'],
                                   latitudes=df2['Latitude'],
                                   longitudes=df2['Longitude']
                                  )


In [19]:
print(try_venues.shape) #venues near Toronto latitude and longitude coordinates (radius 500 m)
try_venues.head()

(309, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0       Domlur              12.973291               77.640467   
1       Domlur              12.973291               77.640467   
2       Domlur              12.973291               77.640467   
3       Domlur              12.973291               77.640467   
4       Domlur              12.973291               77.640467   

                    Venue  Venue Latitude  Venue Longitude    Venue Category  
0               Vero Moda       12.972808        77.641225          Boutique  
1           Chakum Chukum       12.972516        77.639152       Snack Place  
2          Sri Udupi Park       12.973617        77.641104  Udupi Restaurant  
3  Krispy Kreme Doughnuts       12.970094        77.640671            Bakery  
4        The Black Rabbit       12.969891        77.641251               Pub

In [20]:
try_venues.head(20)

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0              Domlur              12.973291               77.640467   
1              Domlur              12.973291               77.640467   
2              Domlur              12.973291               77.640467   
3              Domlur              12.973291               77.640467   
4              Domlur              12.973291               77.640467   
5              Domlur              12.973291               77.640467   
6              Domlur              12.973291               77.640467   
7              Domlur              12.973291               77.640467   
8              Domlur              12.973291               77.640467   
9              Domlur              12.973291               77.640467   
10        Indiranagar              13.016341               77.558664   
11        Indiranagar              13.016341               77.558664   
12        Indiranagar              13.016341               77.558664   
13        Indiranagar              13.016341               77.558664   
14  Jeevanbheemanagar              12.988234               77.554883   
15  Jeevanbheemanagar              12.988234               77.554883   
16  Jeevanbheemanagar              12.988234               77.554883   
17  Jeevanbheemanagar              12.988234               77.554883   
18  Jeevanbheemanagar              12.988234               77.554883   
19  Jeevanbheemanagar              12.988234               77.554883   

                                     Venue  Venue Latitude  Venue Longitude  \
0                                Vero Moda       12.972808        77.641225   
1                            Chakum Chukum       12.972516        77.639152   
2                           Sri Udupi Park       12.973617        77.641104   
3                   Krispy Kreme Doughnuts       12.970094        77.640671   
4                         The Black Rabbit       12.969891        77.641251   
5                Smoor Chocolates by Bliss       12.969913        77.641283   
6                             Corner House       12.969973        77.637453   
7   The Humming Tree - Music Venue and Bar       12.970125        77.639161   
8                                   Stoner       12.976363        77.641188   
9                                 Mamagoto       12.970113        77.639173   
10                                  Durbar       13.012910        77.555740   
11                         Cafe Coffee Day       13.014264        77.555670   
12                    Metro Cash and Carry       13.014283        77.555214   
13                            Juice Center       13.017341        77.562790   
14                      O.G. Variar & Sons       12.991578        77.554561   
15                               Vade Shop       12.986324        77.555936   
16                               Guru Arts       12.988665        77.552434   
17                          Sonu Dahi Puri       12.990884        77.556162   
18                    Navrang Kabab Corner       12.990862        77.558043   
19                 Yamaha-Bangalore Wheels       12.988922        77.559292   

         Venue Category  
0              Boutique  
1           Snack Place  
2      Udupi Restaurant  
3                Bakery  
4                   Pub  
5        Chocolate Shop  
6        Ice Cream Shop  
7           Music Venue  
8          Dessert Shop  
9      Asian Restaurant  
10                  Bar  
11          Coffee Shop  
12     Department Store  
13            Juice Bar  
14               Bakery  
15          Snack Place  
16  Arts & Crafts Store  
17          Snack Place  
18    Indian Restaurant  
19      Motorcycle Shop

## Analyse the Area

In [21]:
try_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                
Anjanapura                               4                       4      4   
Arekere                                 10                      10     10   
BTM Layout                               5                       5      5   
Banashankari                             9                       9      9   
Banaswadi                                4                       4      4   
Basavanagudi                             5                       5      5   
Basaveshwaranagar                        5                       5      5   
Begur                                    2                       2      2   
Bellandur                                1                       1      1   
Bommasandra                             10                      10     10   
CV Raman Nagar                          10                      10     10   
Domlur                                  10                      10     10   
Electronic City                         10                      10     10   
Girinagar                                5                       5      5   
Gottigere                                4                       4      4   
HBR Layout                               4                       4      4   
HSR Layout                               4                       4      4   
Hebbal                                   7                       7      7   
Hoodi                                    3                       3      3   
Horamavu                                 4                       4      4   
Hulimavu                                 4                       4      4   
Indiranagar                              4                       4      4   
J. P. Nagar                              1                       1      1   
Jalahalli                                4                       4      4   
Jayanagar                                4                       4      4   
Jeevanbheemanagar                        6                       6      6   
Kamakshipalya                            2                       2      2   
Kammanahalli                             4                       4      4   
Kengeri                                 10                      10     10   
Koramangala                             10                      10     10   
Kothnur                                  4                       4      4   
Krishnarajapuram                         8                       8      8   
Kumaraswamy Layout                       3                       3      3   
Lingarajapuram                           8                       8      8   
Madiwala                                10                      10     10   
Mahadevapura                             8                       8      8   
Mahalakshmi Layout                       3                       3      3   
Malleswaram                             10                      10     10   
Marathahalli                             4                       4      4   
Mathikere                                2                       2      2   
Pete area                               10                      10     10   
R. T. Nagar                              1                       1      1   
Rajajinagar                             10                      10     10   
Ramamurthy Nagar                         4                       4      4   
Sadashivanagar                           3                       3      3   
Seshadripuram                           10                      10     10   
Shivajinagar                             3                       3      3   
Ulsoor                                   5                       5      5   
Uttarahalli                              5                       5      5   
Varthur                                  3                       3      3   
Vasanth Nagar    

In [22]:
print('There are {} uniques categories.'.format(len(try_venues['Venue Category'].unique())))

There are 102 uniques categories.


Neighborhood  Accessories Store  Andhra Restaurant  Art Gallery  Art Museum  \
0       Domlur                  0                  0            0           0   
1       Domlur                  0                  0            0           0   
2       Domlur                  0                  0            0           0   
3       Domlur                  0                  0            0           0   
4       Domlur                  0                  0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  BBQ Joint  \
0                    0                 0                   0          0   
1                    0                 0                   0          1   
2                    0                 0                   0          0   
3                    0                 0                   0          0   
4                    0                 0                   0          0   

   Badminton Court  Bakery  Bar  Bookstore  Boutique  Breakfast Spot  \
0                0       0    0          0         0               0   
1                0       0    0          0         0               0   
2                0       0    0          0         0               0   
3                0       0    0          0         0               0   
4                0       0    0          0         0               0   

   Burger Joint  Bus Line  Bus Station  Bus Stop  Café  Capitol Building  \
0             0         0            0         0     1                 0   
1             0         0            0         0     0                 0   
2             0         0            0         0     0                 0   
3             0         0            0         0     0                 0   
4             0         0            0         0     0                 0   

   Chettinad Restaurant  Chinese Restaurant  Chocolate Shop  Clothing Store  \
0                     0                   0               0               0   
1                     0                   0               0               0   
2                     0                   0               0               0   
3                     0                   0               0               0   
4                     0                   0               0               0   

   Coffee Shop  Convenience Store  Deli / Bodega  Department Store  \
0            0                  0              0                 0   
1            0                  0              0                 0   
2            0                  0              1                 0   
3            0                  0              0                 0   
4            0                  0              0                 0   

   Dessert Shop  Dim Sum Restaurant  Diner  Donut Shop  \
0             0                   0      0           0   
1             0                   0      0           0   
2             0                   0      0           0   
3             0                   0      0           0   
4             0                   0      0           0   

   Eastern European Restaurant  Electronics Store  Fast Food Restaurant  \
0                            0                  0                     0   
1                            0                  0                     0   
2                            0                  0                     0   
3                            0                  0                     0   
4                            0                  0                     0   

   Flea Market  Food & Drink Shop  Food Court  Food Truck  \
0            0                  0           0           0   
1            0                  0           0           0   
2            0                  0           0           0   
3            0                  0           0           0   
4            0                  0           0           0   

   Furniture / Home Store  Garden  Golf Course  Gourmet Shop  Grocery Store  \
0                       0       

In [23]:
bengaluru_onehot = pd.get_dummies(try_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
bengaluru_onehot['Neighborhood'] = try_venues['Neighborhood'] 

# move neighborhood column to the first column
#fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
col_list = list(bengaluru_onehot.columns)
col_list.remove('Neighborhood')
fixed_columns = ['Neighborhood'] + col_list
bengaluru_onehot = bengaluru_onehot[fixed_columns]

bengaluru_onehot.head()

Neighborhood  Accessories Store  Andhra Restaurant  Art Gallery  Art Museum  \
0       Domlur                  0                  0            0           0   
1       Domlur                  0                  0            0           0   
2       Domlur                  0                  0            0           0   
3       Domlur                  0                  0            0           0   
4       Domlur                  0                  0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  BBQ Joint  \
0                    0                 0                   0          0   
1                    0                 0                   0          0   
2                    0                 0                   0          0   
3                    0                 0                   0          0   
4                    0                 0                   0          0   

   Badminton Court  Bakery  Bar  Bookstore  Boutique  Breakfast Spot  \
0                0       0    0          0         1               0   
1                0       0    0          0         0               0   
2                0       0    0          0         0               0   
3                0       1    0          0         0               0   
4                0       0    0          0         0               0   

   Burger Joint  Bus Line  Bus Station  Bus Stop  Café  Capitol Building  \
0             0         0            0         0     0                 0   
1             0         0            0         0     0                 0   
2             0         0            0         0     0                 0   
3             0         0            0         0     0                 0   
4             0         0            0         0     0                 0   

   Chettinad Restaurant  Chinese Restaurant  Chocolate Shop  Clothing Store  \
0                     0                   0               0               0   
1                     0                   0               0               0   
2                     0                   0               0               0   
3                     0                   0               0               0   
4                     0                   0               0               0   

   Coffee Shop  Convenience Store  Department Store  Dessert Shop  \
0            0                  0                 0             0   
1            0                  0                 0             0   
2            0                  0                 0             0   
3            0                  0                 0             0   
4            0                  0                 0             0   

   Dim Sum Restaurant  Diner  Donut Shop  Eastern European Restaurant  \
0                   0      0           0                            0   
1                   0      0           0                            0   
2                   0      0           0                            0   
3                   0      0           0                            0   
4                   0      0           0                            0   

   Electronics Store  Fast Food Restaurant  Flea Market  Food & Drink Shop  \
0                  0                     0            0                  0   
1                  0                     0            0                  0   
2                  0                     0            0                  0   
3                  0                     0            0                  0   
4                  0                     0            0                  0   

   Food Court  Food Truck  Furniture / Home Store  Garden  Golf Course  \
0           0           0                       0       0            0   
1           0           0                       0       0            0   
2           0           0                       0       0            0   
3           0           0                       0       0            0   
4     

In [24]:
bengaluru_grouped = bengaluru_onehot.groupby('Neighborhood').mean().reset_index()
bengaluru_grouped

Neighborhood  Accessories Store  Andhra Restaurant  Art Gallery  \
0           Anjanapura                 0.0               0.00          0.0   
1              Arekere                 0.0               0.00          0.0   
2           BTM Layout                 0.0               0.00          0.0   
3         Banashankari                 0.0               0.00          0.0   
4            Banaswadi                 0.0               0.25          0.0   
5         Basavanagudi                 0.2               0.00          0.0   
6    Basaveshwaranagar                 0.0               0.00          0.0   
7                Begur                 0.0               0.00          0.0   
8             Bellandur                0.0               0.00          0.0   
9          Bommasandra                 0.0               0.00          0.0   
10       CV Raman Nagar                0.0               0.00          0.0   
11               Domlur                0.0               0.00          0.0   
12     Electronic City                 0.0               0.00          0.0   
13           Girinagar                 0.0               0.00          0.0   
14           Gottigere                 0.0               0.00          0.0   
15          HBR Layout                 0.0               0.00          0.0   
16          HSR Layout                 0.0               0.00          0.0   
17              Hebbal                 0.0               0.00          0.0   
18                Hoodi                0.0               0.00          0.0   
19            Horamavu                 0.0               0.00          0.0   
20            Hulimavu                 0.0               0.00          0.0   
21          Indiranagar                0.0               0.00          0.0   
22         J. P. Nagar                 0.0               0.00          0.0   
23           Jalahalli                 0.0               0.00          0.0   
24           Jayanagar                 0.0               0.00          0.0   
25    Jeevanbheemanagar                0.0               0.00          0.0   
26       Kamakshipalya                 0.0               0.00          0.0   
27        Kammanahalli                 0.0               0.00          0.0   
28             Kengeri                 0.0               0.00          0.0   
29         Koramangala                 0.0               0.00          0.0   
30              Kothnur                0.0               0.00          0.0   
31     Krishnarajapuram                0.0               0.00          0.0   
32  Kumaraswamy Layout                 0.0               0.00          0.0   
33      Lingarajapuram                 0.0               0.00          0.0   
34            Madiwala                 0.0               0.00          0.0   
35         Mahadevapura                0.0               0.00          0.0   
36  Mahalakshmi Layout                 0.0               0.00          0.0   
37          Malleswaram                0.0               0.00          0.0   
38         Marathahalli                0.0               0.00          0.0   
39           Mathikere                 0.0               0.00          0.0   
40            Pete area                0.0               0.00          0.0   
41         R. T. Nagar                 0.0               0.00          0.0   
42          Rajajinagar                0.0               0.00          0.0   
43    Ramamurthy Nagar                 0.0               0.00          0.0   
44       Sadashivanagar                0.0               0.00          0.0   
45        Seshadripuram                0.0               0.00          0.1   
46         Shivajinagar                0.0               0.00          0.0   
47               Ulsoor                0.0               0.00          0.0   
48          Uttarahalli                0.0               0.00          0.0   
49              Varthur                0.0               0.00          0.0   
50        Vasanth Nagar          

### Fetch Top 5 frequently visting venus

In [25]:
num_top_venues = 5

for hood in bengaluru_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = bengaluru_grouped[bengaluru_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')



----Anjanapura ----
                           venue  freq
0               Department Store  0.25
1                      Juice Bar  0.25
2  Vegetarian / Vegan Restaurant  0.25
3             Athletics & Sports  0.25
4                  Metro Station  0.00


----Arekere ----
                  venue  freq
0     Indian Restaurant   0.2
1  Fast Food Restaurant   0.2
2          Burger Joint   0.1
3             Juice Bar   0.1
4  Karnataka Restaurant   0.1


----BTM Layout ----
                  venue  freq
0         Metro Station   0.2
1     Indian Restaurant   0.2
2             Juice Bar   0.2
3  Gym / Fitness Center   0.2
4          Liquor Store   0.2


----Banashankari ----
               venue  freq
0     Sandwich Place  0.22
1               Café  0.22
2       Burger Joint  0.11
3  Indian Restaurant  0.11
4     Ice Cream Shop  0.11


----Banaswadi ----
               venue  freq
0       Burger Joint  0.25
1  Andhra Restaurant  0.25
2             Bakery  0.25
3               Café  0.25
4  

### Listing the most commonly visting venues in an Area

In [26]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [27]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = bengaluru_grouped['Neighborhood']

for ind in np.arange(bengaluru_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(bengaluru_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()


Neighborhood 1st Most Common Venue          2nd Most Common Venue  \
0    Anjanapura       Department Store  Vegetarian / Vegan Restaurant   
1       Arekere      Indian Restaurant           Fast Food Restaurant   
2    BTM Layout      Indian Restaurant           Gym / Fitness Center   
3  Banashankari                   Café                 Sandwich Place   
4     Banaswadi           Burger Joint              Andhra Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0             Juice Bar    Athletics & Sports            Food Truck   
1        Ice Cream Shop             Juice Bar           Supermarket   
2             Juice Bar          Liquor Store         Metro Station   
3     Indian Restaurant        Breakfast Spot            Food Truck   
4                  Café                Bakery         Women's Store   

  6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0          Dessert Shop    Dim Sum Restaurant                        Diner   
1  Karnataka Restaurant           Snack Place                 Burger Joint   
2     Food & Drink Shop          Dessert Shop           Dim Sum Restaurant   
3          Burger Joint        Ice Cream Shop           Athletics & Sports   
4                 Diner            Donut Shop  Eastern European Restaurant   

  9th Most Common Venue       10th Most Common Venue  
0            Donut Shop  Eastern European Restaurant  
1            Food Court           Dim Sum Restaurant  
2                 Diner                   Donut Shop  
3             BBQ Joint                        Diner  
4     Electronics Store         Fast Food Restaurant

In [63]:


# set number of clusters
kclusters = 10

bengaluru_grouped_clustering = bengaluru_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(bengaluru_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]



array([1, 1, 1, 5, 5, 2, 2, 6, 8, 3], dtype=int32)

In [64]:
kmeans.labels_

array([1, 1, 1, 5, 5, 2, 2, 6, 8, 3, 3, 2, 3, 5, 2, 2, 1, 1, 3, 5, 3, 1,
       4, 2, 5, 2, 2, 5, 2, 3, 5, 5, 2, 5, 1, 1, 2, 5, 5, 5, 1, 7, 3, 5,
       3, 1, 0, 3, 5, 3, 3, 5, 9, 0, 2], dtype=int32)

In [65]:
result_latitude = []
result_longitude = []
for index, row in neighborhoods_venues_sorted.iterrows():
    
    neigh = row.Neighborhood
    x = df.loc[df['Neighborhood'] == neigh]
    latitude = x.iloc[0]['Latitude']
    result_latitude.append(latitude)
    longitude = x.iloc[0]['Longitude']
    result_longitude.append(longitude)

neighborhoods_venues_sorted["Latitude"] = pd.Series(result_latitude)
neighborhoods_venues_sorted["Longitude"] = pd.Series(result_longitude)

In [66]:
neighborhoods_venues_sorted['Cluster Labels'] = kmeans.labels_

In [67]:
neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue          2nd Most Common Venue  \
0    Anjanapura       Department Store  Vegetarian / Vegan Restaurant   
1       Arekere      Indian Restaurant           Fast Food Restaurant   
2    BTM Layout      Indian Restaurant           Gym / Fitness Center   
3  Banashankari                   Café                 Sandwich Place   
4     Banaswadi           Burger Joint              Andhra Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0             Juice Bar    Athletics & Sports            Food Truck   
1        Ice Cream Shop             Juice Bar           Supermarket   
2             Juice Bar          Liquor Store         Metro Station   
3     Indian Restaurant        Breakfast Spot            Food Truck   
4                  Café                Bakery         Women's Store   

  6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0          Dessert Shop    Dim Sum Restaurant                        Diner   
1  Karnataka Restaurant           Snack Place                 Burger Joint   
2     Food & Drink Shop          Dessert Shop           Dim Sum Restaurant   
3          Burger Joint        Ice Cream Shop           Athletics & Sports   
4                 Diner            Donut Shop  Eastern European Restaurant   

  9th Most Common Venue       10th Most Common Venue   Latitude  Longitude  \
0            Donut Shop  Eastern European Restaurant  12.874255  77.583725   
1            Food Court           Dim Sum Restaurant  12.986475  77.538571   
2                 Diner                   Donut Shop  12.915221  77.573598   
3             BBQ Joint                        Diner  12.906817  77.563525   
4     Electronics Store         Fast Food Restaurant  13.011768  77.627829   

   Cluster Labels  
0               1  
1               1  
2               1  
3               5  
4               5

In [85]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(neighborhoods_venues_sorted['Latitude'], neighborhoods_venues_sorted['Longitude'], neighborhoods_venues_sorted['Neighborhood'], neighborhoods_venues_sorted['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters



# Filtering the data and thus suggesting the Places where Indian restaurant can be opened

In [82]:
mylist=neighborhoods_venues_sorted['1st Most Common Venue']
sub = 'Indian Restaurant'
listing=[sub in mystring for mystring in mylist]
ind_restaurant_details=neighborhoods_venues_sorted[listing].reset_index(drop=True)
ind_restaurant_details.head(12)

Neighborhood 1st Most Common Venue          2nd Most Common Venue  \
0           Arekere      Indian Restaurant           Fast Food Restaurant   
1        BTM Layout      Indian Restaurant           Gym / Fitness Center   
2       Bommasandra      Indian Restaurant                 Sandwich Place   
3     CV Raman Nagar     Indian Restaurant                  Movie Theater   
4   Electronic City      Indian Restaurant              Indian Sweet Shop   
5        HSR Layout      Indian Restaurant                      Juice Bar   
6            Hebbal      Indian Restaurant             Chinese Restaurant   
7              Hoodi     Indian Restaurant               Department Store   
8          Horamavu      Indian Restaurant                          Hotel   
9          Hulimavu      Indian Restaurant                         Temple   
10     Kammanahalli      Indian Restaurant  Vegetarian / Vegan Restaurant   
11      Koramangala      Indian Restaurant                          Diner   

   3rd Most Common Venue   4th Most Common Venue 5th Most Common Venue  \
0         Ice Cream Shop               Juice Bar           Supermarket   
1              Juice Bar            Liquor Store         Metro Station   
2    Indie Movie Theater  Furniture / Home Store  Fast Food Restaurant   
3    Sporting Goods Shop       Kerala Restaurant                Lounge   
4             Hookah Bar                Tea Room                  Café   
5             Smoke Shop                    Park     Food & Drink Shop   
6           Dessert Shop        Department Store           Coffee Shop   
7               Pharmacy              Food Court          Dessert Shop   
8                 Bakery                     Bar            Food Truck   
9       Asian Restaurant                Bus Line            Food Court   
10                Bakery           Shopping Mall            Food Truck   
11    Italian Restaurant                     Gym                Bakery   

   6th Most Common Venue          7th Most Common Venue 8th Most Common Venue  \
0   Karnataka Restaurant                    Snack Place          Burger Joint   
1      Food & Drink Shop                   Dessert Shop    Dim Sum Restaurant   
2       Department Store                       Bus Stop        Ice Cream Shop   
3         Clothing Store  Vegetarian / Vegan Restaurant   Arts & Crafts Store   
4             Restaurant       Mediterranean Restaurant           Flea Market   
5       Department Store                   Dessert Shop    Dim Sum Restaurant   
6            Pizza Place                 Ice Cream Shop     Andhra Restaurant   
7     Dim Sum Restaurant                          Diner            Donut Shop   
8     Dim Sum Restaurant                          Diner            Donut Shop   
9           Dessert Shop             Dim Sum Restaurant                 Diner   
10          Dessert Shop             Dim Sum Restaurant                 Diner   
11                Garden       Mediterranean Restaurant            Food Court   

          9th Most Common Venue       10th Most Common Venue   Latitude  \
0                    Food Court           Dim Sum Restaurant  12.986475   
1                         Diner                   Donut Shop  12.915221   
2                     BBQ Joint              Badminton Court  12.921227   
3                    Food Court                        Diner  12.955257   
4                  Dessert Shop           Dim Sum Restaurant  12.941726   
5                         Diner                   Donut Shop  12.940144   
6                         Diner                   Donut Shop  13.076641   
7   Eastern European Restaurant            Electronics Store  12.940615   
8   Eastern European Restaurant            Electronics Store  13.038218   
9                    Donut Shop  Eastern European Restaurant  13.011894   
10                   Donut Shop  Eastern European Restaurant  13.046453   
11           Dim Sum Restaurant                   Donut Shop  12.909492   

    Longitude

# Filtering the thus suggesting the place(s) for opening up of Cafe

In [84]:
mylist=neighborhoods_venues_sorted['1st Most Common Venue']
sub = 'Café'
listing=[sub in mystring for mystring in mylist]
ind_cafe=neighborhoods_venues_sorted[listing].reset_index(drop=True)
ind_cafe.head(12)

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Banashankari                   Café        Sandwich Place   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0     Indian Restaurant        Breakfast Spot            Food Truck   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0          Burger Joint        Ice Cream Shop    Athletics & Sports   

  9th Most Common Venue 10th Most Common Venue   Latitude  Longitude  \
0             BBQ Joint                  Diner  12.906817  77.563525   

   Cluster Labels  
0               5

In [69]:
neighborhoods_venues_sorted.columns.values

array(['Neighborhood', '1st Most Common Venue', '2nd Most Common Venue',
       '3rd Most Common Venue', '4th Most Common Venue',
       '5th Most Common Venue', '6th Most Common Venue',
       '7th Most Common Venue', '8th Most Common Venue',
       '9th Most Common Venue', '10th Most Common Venue', 'Latitude',
       'Longitude', 'Cluster Labels'], dtype=object)